tools/c7n_azure/c7n_azure/resources/storage.py

```diff
@@ -246,33 +244,102 @@ def process_resource_set(self, resources, event=None, session=None, token=None):
+            storage_prefix_property = get_annotation_prefix(BLOB_TYPE)
+
+        elif self.storage_type == FILE_TYPE:
+            storage_prefix_property = get_annotation_prefix(FILE_TYPE)
+
+        elif self.storage_type == TABLE_TYPE:
+            storage_prefix_property = get_annotation_prefix(TABLE_TYPE)
+
+        elif self.storage_type == QUEUE_TYPE:
+            storage_prefix_property = get_annotation_prefix(QUEUE_TYPE)
+
+        if not (storage_prefix_property in storage_account):
+            settings = StorageSettingsUtilities.get_settings(
+                self.storage_type, storage_account, session, token)
+            storage_account[storage_prefix_property] = json.loads(jsonpickle.encode(settings))
+
+        return storage_account[storage_prefix_property]
+
+
+@Storage.action_registry.register('update-log-settings')
+class UpdateLogSettingsAction(AzureBaseAction):
+    """Action that updates the logging settings on storage accounts. The action requires
+    specifying an array of storage types that will be impacted by the action (blob, queue, table),
+    retention (number in days; 0-365), and an array of log settings to enable (read, write, delete).
+    The action will disable any settings not listed (e.g. by providing log: [write, delete], the
+    action will disable read).
+
+     :example:
+
+        Enable write and delete logging and disable read logging on blob storage,
+        and retain logs for 5 days.
+
+     .. code-block:: yaml
+
+        policies:
+            - name: enable-blob-storage-logging
+              resource: azure.storage
+              actions:
+                - type: update-log-settings
+                  storage-types: [blob]
+                  retention: 5
+                  log: [write, delete]
+    """
+
+    READ = 'read'
+    WRITE = 'write'
+    DELETE = 'delete'
+
+    schema = type_schema('update-log-settings',
+                         required=['storage-types', 'log', 'retention'],
+                         **{
+                             'storage-types': {
+                                 'type': 'array',
+                                 'items': {
+                                     'type': 'string',
+                                     'enum': [BLOB_TYPE, QUEUE_TYPE, TABLE_TYPE]
+                                 }
+                             },
+                             'log': {
+                                 'type': 'array',
+                                 'items': {
+                                     'type': 'string',
+                                     'enum': [READ, WRITE, DELETE]
+                                 }
+                             },
+                             'retention': {'type': 'number'}
+                         }
+                         )
+
+    def __init__(self, data, manager=None):
+        super(UpdateLogSettingsAction, self).__init__(data, manager)
+        self.storage_types = data.get('storage-types')
```

Comment:
nit: you can use `[]` instead of `.get()` for required fields.. Will fail fast early if for some reason schema is incorrect or schema validator has a bug